### ISSUE TITLE
Theme notification is send twice on some nodes

### ISSUE THREAD

**Issue by Maran23** [godotengine/godot#70416]
### Godot version

4.x

### System information

Windows 10

### Issue description

Some nodes receive the theme notification twice.
The problematic code:
![image](https://user-images.githubusercontent.com/66004280/208978140-ce8065aa-8562-4d0a-bbe9-c9b3c15019e5.png)

The `gui_base` can be a direct child of `theme_base`. 
In this case the `theme_base` sets the theme and propagates the notification to all children down in the scene tree, including `gui_base`.
Then `gui_base` does the same again.

### Steps to reproduce

Found when testing https://github.com/godotengine/godot/pull/70410

1. Add a breakpoint to the `NOTIFICATION_THEME_CHANGED` in e.g. `editor_audio_buses.cpp`
![image](https://user-images.githubusercontent.com/66004280/208989313-32a2061a-da85-495a-87a3-cad28aa0bdd7.png)
2. Change the editor theme
3. See how `NOTIFICATION_THEME_CHANGED` is called twice

In this case the following condition is true: `gui_base->get_parent() == theme_base`. There are probably more combinations out there...

### Minimal reproduction project

--

**Comment by Maran23** [comment]
cc @YuriSizov - Not a big problem but maybe still worth checking out :)

**Comment by YuriSizov** [comment]
We plan to refactor EditorNode in 4.1+, and this would also go. I think @AaronRecord had a PR that simplified this theme situation for 4.0, but we didn't merge it yet.

**Comment by Maran23** [comment]
Good idea. I also saw https://github.com/godotengine/godot/pull/62846, which unfortunately was reverted due to problems. 
But this is also a good refactoring. 
While testing, I also found that `NOTIFICATION_ENTER_TREE ` paired with `NOTIFICATION_THEME_CHANGED` is almost never needed and nothing broke when I tried removing it. So definitely worth doing on some point.

**Comment by YuriSizov** [comment]
This has been improved in the majority of GUI classes, but still needs to be done in the editor code. We now have a virtual method that can be implemented on a Control/Window derivative that you'd use to cache all theme properties as needed. Most of built-in controls no longer need to bind to either notification. Some still do because I had to omit them from my PRs as they've been refactored by other PRs in the process.

See https://github.com/godotengine/godot/pull/65156 and https://github.com/godotengine/godot/pull/65250.

Same changes need to be done to the editor code, where we subscribe to theme changes. It needs to be moved to use the new method and the theme_cache pattern.
